### ISSUE TITLE
Build and install downstream packages against pre-installed Kokkos built with TriBITS

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#11545]
## Description

???

**Comment by ibaned** [comment]
Thanks for working on this! I will greatly help our Sandia production code when this is done. I'd be happy to help with making the native Kokkos build system export everything TriBITS expects.

**Comment by bartlettroscoe** [comment]
CC: @crtrott, @rppawlo, @ndellingwood, @ibaned, @sebrowne

FYI: The following PR removes TriBITS subpackages from the TriBITS build of Kokkos and downstream Trilinos packages:

* #11808 

Merging this PR opens the door to getting Trilinos to work with the native Kokkos CMake build system.

**Comment by bartlettroscoe** [comment]
FYI: I have branches for Trilinos and TriBITS where Trilinos is working with Kokkos (and KokkosConfig.cmake) installed using the native Kokkos CMake build system.  It requires only very minor changes to the Kokkos native CMake build system and some more significant changes to TriBITS (mostly documentation 😌).  Given that this requires some tweaks to TriBITS, I was going to do a TriBITS snapshot update into Trilinos 'develop' first, and then rebase the branch in PR #11808 on top of that.  I can then push the remaining changes to Kokkos to that branch/PR or open a new [stacked PR](https://blog.logrocket.com/using-stacked-pull-requests-in-github/) against PR #11808.   I am leaning on just pushing the updated commits to PR PR #11808 and expanding the scope of that PR (which is just a few more small commits).

**Comment by bartlettroscoe** [comment]
CC: @crtrott, @rppawlo, @ndellingwood, @ibaned, @sebrowne, @jwillenbring, @jhux2

> FYI: I have branches for Trilinos and TriBITS where Trilinos is working with Kokkos (and KokkosConfig.cmake) installed using the native Kokkos CMake build system. 

PR:

* https://github.com/trilinos/Trilinos/pull/11863

contains these changes to Kokkos.  The associated TriBITS changes are in:

* https://github.com/TriBITSPub/TriBITS/pull/576

Just a few things to do before merging that TriBITS PR TriBITSPub/TriBITS#576 and then these change can be pretty quickly merged to Trilinos 'develop' and Kokkos 'develop' and the issue of pre-installing Kokkos using its native CMake build system and allowing remaining Trilinos packages to build against that will be solved once and for all.  We should be able to button that up in the next few days (assuming people can do reviews and approve PRs so they can merge).

Currently, the biggest impediment to all of this is getting the PR:

* #11808 

merged to Trilinos 'develop'.  Those changes have already been merged to Kokkos and KokosKerenls 'develop' in their own repos.

**Comment by bartlettroscoe** [comment]
FYI: PR:

* #11863 

should be ready to approve and merge.

Once that PR is merged, I can post another PR to some final cleanup (see [here](https://github.com/trilinos/Trilinos/pull/11808#discussion_r1176655469)) and this should be done.

**Comment by bartlettroscoe** [comment]
FYI: There were some issues with the removal of subpackages from Kokkos with the merge of PR #11808 that had to be addressed:

* #11938
  * Addressed in commit https://github.com/trilinos/Trilinos/pull/11863/commits/6bffd4453c77f5f96c3db4eb0d4dcb6644c14f66 from PR #11863
* #11940
  * Addressed in commit https://github.com/trilinos/Trilinos/pull/11863/commits/59c2e4d52b9bf73563f13752a8175735eb4cc64a in PR #11863

Hopefully that

**Comment by bartlettroscoe** [comment]
I sent the following email response that gives a good summary of the current status of this use case.

----

From: Bartlett, Roscoe A 
Sent: Monday, June 5, 2023 12:08 PM
To: `<...>`
Cc: `<...>`
Subject: RE: External Kokkos in Trilinos build

...

> Sameer was asking if the capability to pull in Kokkos as an external package is ready to deploy in the E4S build. I thought the most recent status was that it is possible to build Kokkos separately and then pull it into building Trilinos, but that building Kokkos with its native build system (that would be used by Spack) and then pulling that in was not quite ready yet.  Is that accurate?

That is not accurate.  That use case is fully supported as of the Trilinos PR:

*   https://github.com/trilinos/Trilinos/pull/11863

(see the demonstrations with [clang](https://github.com/trilinos/Trilinos/pull/11863#issuecomment-1569271418) and [cuda](https://github.com/trilinos/Trilinos/pull/11863#issuecomment-1572386063) in that PR).  I just need someone to approve that PR and it will be merged to Trilinos ‘develop’ (as it has passed all of the PR builds and tests).

With the patched version of Kokkos 4.0.1 in the above Trilinos PR (which will be on Trilinos ‘develop’ as soon as the above PR is approved so it can be merged), that use case should be fully supported.  The matching Kokkos changes have already been merged into Kokkos ‘develop’ (see https://github.com/kokkos/kokkos/pull/6142, https://github.com/kokkos/kokkos/pull/6157, and https://github.com/kokkos/kokkos/pull/6176).

**HOWEVER: The problem is that you can’t currently build Trilinos ‘develop’ against Kokkos ‘develop’ (as Kokkos ‘develop’ has broken backward compatibility w.r.t. Trilinos since Kokkos 4.0.1 was released and snapshotted into Trilinos ‘develop’).** (See [here](https://github.com/trilinos/Trilinos/pull/11863#issuecomment-1570559985).)

Therefore, to take advantage of this in Spack with Trilinos ‘develop’, one of the following must happen on the Kokkos side:

* **Option-a**: Create a new Kokkos patch release off of Kokkos 4.0.1 (Kokkos 4.0.2?) that pulls in the needed Kokkos CMake changes.  (That would technically violate semver.org but it should maintain backward compatibility for users so it should be okay in this case.)
* **Option-b**: Have Kokkos put out a new special release called something like `4.0.1-tril-dev` that matches the exact version of Kokkos in Trilinos ‘develop’ (as of the merge of the above Trilinos PR #11863).
* **Option-c**: Wait for Kokkos to put out its next release taken off of the Kokkos ‘develop’ branch (Kokkos 4.1.0?)

Option-a and option-b could be done very quickly.  Option-c will take longer as it will need to wait for the next Kokkos release.

After that, the Spack trilinos/package.py file can be updated with the configure option `-DTPL_ENABLE_Kokkos=ON` (see [here](https://docs.trilinos.org/files/TrilinosBuildReference.html#building-against-pre-installed-packages)) and the dependency on the most recent version of Kokkos (as per above) can be specified as a dependency for Trilinos ‘develop’ (and that should be the only changes needed as Spack and CMake will take care of the rest of the details).  And with that,  Spack users will be able to build with trilinos@develop against the sperate Spack Kokkos package.

However, if one wants to use official releases of Trilinos and Kokkos (instead of Trilinos ‘develop’ and a special Kokkos release), then the following will need to occur:

1)	Kokkos needs to put out the next release off of Kokkos ‘develop’ (i.e. Kokkos 4.1.0?)
2)	The version of Kokkos in Trilinos ‘develop’ needs to be updated to the new Kokkos release
3)	Trilinos needs to put on release Trilinos 4.1.0 off of Trilinos ‘develop’
4)	The Spack package file trilinos/package.py needs to be updated as described above (i.e. configure with `-DTPL_ENABLE_Kokkos=ON` and depend on [kokkos@4.1.0](mailto:kokkos@4.1.0))

At that point, Spack users can build with [trilinos@14.1.0](mailto:trilinos@14.1.0) and it will use the separate [kokkos@4.1.0](mailto:kokkos@4.1.0) package.  But that will take several months to occur.

`<... Kokkos Developers ...>`,

If I have missed anything, please let me know.

All,

If there is anything more that I can do to make this happen, please let me know (I think that I have done all that I can do at this point).  For starters, we need someone to approve Trilinos PR [#11863](https://github.com/trilinos/Trilinos/pull/11863) so it can be merged to Trilinos ‘develop’.

Thanks,

-Ross

**Comment by bartlettroscoe** [comment]
FYI: Here is the Spack PR that has the Trilinos Spack package using the Kokkos Spack package:

* https://github.com/spack/spack/pull/39397

**Comment by sebrowne** [comment]
I agree with this approach.  I had no idea that worked (`find_package(Tpetra)`, for instance), but that's excellent!

**Comment by bartlettroscoe** [comment]
> I agree with this approach. I had no idea that worked (`find_package(Tpetra)`, for instance), but that's excellent!

@sebrowne, this has been supported for a least 10 years (support for which was not written by me).   I have added more explicit documentation for this as part of this TriBITS/Trilinos CMake modernization work.  For example, see:

* https://docs.trilinos.org/files/TrilinosBuildReference.html#using-the-installed-software-in-downstream-cmake-projects

and see a (tested) example specific to Trilinos at:

* https://github.com/trilinos/Trilinos/tree/develop/demos/simpleBuildAgainstTrilinos#3-configure-the-example-project
* https://github.com/trilinos/Trilinos/blob/develop/demos/simpleBuildAgainstTrilinos/CMakeLists.by_package.cmake
